sql/src/main/java/io/crate/execution/dml/ShardRequestExecutor.java

```diff
@@ -183,7 +191,35 @@ private static ShardId getShardId(ClusterService clusterService,
         ).shardId();
     }
 
-    private static void updateRowCountOrFail(long[] rowCount, ShardResponse response) {
+    private ActionListener<ShardResponse> rowCountListener(RowConsumer consumer, int numberResponse) {
+        return new MultiActionListener<>(
+            numberResponse,
+            () -> new long[]{0},
+            this::countRows,
+            rowCount -> new Row1(rowCount[0]),
+            new OneRowActionListener<>(consumer, Function.identity())
+        );
+    }
+
+    private ActionListener<ShardResponse> collectValuesListener(RowConsumer consumer, int numberResponse) {
+        return new MultiActionListener<>(
+            numberResponse,
+            ArrayList::new,
+            this::collectValues,
+            result -> new RowN(result.toArray()),
```

Comment:
When there are no return values in the result (an empty array), this will cause an error. How to handle empty results properly here ? `Row.Empty` is also causing problems.